File: 39-combination-sum/combination-sum.java
class Solution {
    void check(List<Integer> s,List<List<Integer>> ans,int i,int[] candidates,int target){
        //SO LAST MAIN ANS MAINKARTHE TIME NEW BANAKE ADD KARNA
        //AND FIRST HAMEASHA PICK KARNA
        if(i==candidates.length) return ;
        if(target==0) {
            ans.add(new ArrayList<Integer>(s));
            return ;
        }
        //pick
        if(candidates[i]<=target){
            s.add(candidates[i]);
            check(s,ans,i,candidates,target-candidates[i]);
            s.remove(s.size()-1);      
        }
        //unpick
        check(s,ans,i+1,candidates,target);
    }
    public List<List<Integer>> combinationSum(int[] candidates, int target) {
        List<List<Integer>> ans=new ArrayList<>();
        List<Integer> s=new ArrayList<>();

        check(s,ans,0,candidates,target);
        return ans;
    }
}

File: 39-combination-sum/combination-sum.py
class Solution(object):
    def combinationSum(self, candidates, target):
        """
        :type candidates: List[int]
        :type target: int
        :rtype: List[List[int]]
        """

        def check(i, tar, subans, ans):
            if tar == 0:
                temp = [i for i in subans]
                ans.append(temp)
                return
            if i == n:
                return 
            
            for j in range(i, n):
                if j != i and candidates[j] == candidates[j - 1]: continue
                if candidates[j] <= tar:
                    subans.append(candidates[j])
                    check(j, tar - candidates[j], subans, ans)
                    subans.pop(-1)
                else:
                    return
        
        candidates.sort()
        n = len(candidates)
        ans = []
        check(0, target, [], ans)
        return ans